pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -122,6 +122,19 @@ func MakeVirtualServices(ia v1alpha1.IngressAccessor, gateways map[v1alpha1.Ingr
 		requiredGatewayCount += gateways[v1alpha1.IngressVisibilityClusterLocal].Len()
 	}
 
+	// CUSTOM LOGIC TO SUPPORT CUSTOM GATEWAYS PER SERVICE.
+	// This is until Knative supports splitting gateways per service as tracked by this PR (https://github.com/knative/serving/pull/4909).
+	customGateway := ia.GetAnnotations()[networking.CustomGatewayAnnotationKey]
+	if customGateway != "" {
+		requiredGatewayCount++
+		// Name format of the custom gateway is either {namespace}/{gateway-name} or {gateway-name}.
+		// If namespace is not specified, it will default to the service namespace.
+		if !strings.Contains(customGateway, "/") {
+			customGateway = VirtualServiceNamespace(ia) + "/" + customGateway
+		}
+		gateways[v1alpha1.IngressVisibilityExternalIP].Insert(customGateway)
```

Comment:
so are we assuming its an external gateway? if we are going to propose to knative, should we have two annotations (one for internal and one for external?)

Reply:
Yes, I'll split this into two.